### ISSUE TITLE
New-style class memory leak

### ISSUE THREAD

**Issue by slide** [IronLanguages/ironpython2#83]
_From @grooya on September 21, 2016 14:47_

New style class objects are holding C# objects that are set as variable. Here is examples of inheritance when using old-style and new-style classes that produce the same result but in first case there is no memory leak while in the second case C# objects are not collected even if garbage collection is forced.
1. Old-style:

```
class Test:
    def __init__(self, variable):
        self.variable = variable

class DerivedTest(Test):
    def __init__(self, *args):
        Test.__init__(self, *args)

testObj = DerivedTest(Variable)
result = testObj.variable.Variable
```
1. New-style:

```
class Test(object):
    def __init__(self, variable):
        self.variable = variable

class DerivedTest(Test):
    def __init__(self, *args):
        super(DerivedTest, self).__init__(*args)

testObj = DerivedTest(Variable)
result = testObj.variable.Variable
```

Note that Variable is set via SetVariable method by instantiating C# Data class.
After running the first script, there are no Data objects in memory, while the second script keeps Data object and it won't release it.

Also, I have performed some measurements when executing this script 10 000 times and got the following results:

Old-style:
Total memory consumption after script execution: 484KB
Total execution time: 231.6ms

New-style:
Total memory consumption after script execution:  1.1MB
Total execution time: 360.6ms

Here is the test console application which presents this issue: [PythonTest.zip](https://github.com/IronLanguages/main/files/485293/PythonTest.zip)

_Copied from original issue: IronLanguages/main#1426_

**Comment by slide** [comment]
_From @simplicbe on September 22, 2016 6:32_

That is very interesting and may explain some issues we had with IronPython in the past. I'll take a close look at this today. Thanks for reporting it with all details!

**Comment by slide** [comment]
_From @vukoje on September 22, 2016 9:26_

@simplicbe it took 2 weeks and 3 guys to pinpoint this issue we are having in production :smile: . I hope this will make a big difference.

**Comment by slide** [comment]
_From @simplicbe on September 22, 2016 9:43_

@grooya that is a very interesting behaviour. Because not all `Data` instances are kept in memory, but a few. The GC root seems to be the `PythonContext`.

**Comment by slide** [comment]
_From @grooya on September 23, 2016 11:59_

I have used dotMemory to track retain graph and got following:

Old style before Data object got collected:
![oldstyle](https://cloud.githubusercontent.com/assets/22348228/18784707/0ded38e2-8194-11e6-8f89-7a3676b4d946.PNG)

New style after all scripts are executed:
![newstyle](https://cloud.githubusercontent.com/assets/22348228/18784726/41168994-8194-11e6-8e36-e28a1e46140e.PNG)

This maybe useful to find part of code which unnecessary holds C# objects even if script execution is done. I hope this could be fixed because "few" instances are causing problems when ironpython is used as part of Web application which execution "never ends" (application pool restart could only release retained memory) .

**Comment by slide** [comment]
_From @simplicbe on September 23, 2016 12:2_

@grooya yep, same result with Teleriks Just Trace :) Thanks for sharing!

**Comment by vukoje** [comment]
Are there any plans to address this issue for 2.7.8?
We are using IronPython to support scripting in cloud platform and memory leaks are making our platform unstable.

Thanks

**Comment by slide** [comment]
I don't think this has been looked into enough that it would make it into 2.7.8. I haven't debugged one of these types of issues in a long time, so unless there is someone else that can step up and look into it as well, it may take some time before it gets some love.

**Comment by slozier** [comment]
For future interested debuggers: I don't think assigning the variable has anything to do with the leak that is demonstrated here. You can easily reproduce the same issue with the following code:
```Python
class Test(object):
    x = [0]*5000
    def __init__(self):
        pass

Test()
```

It looks like the `PythonType` is being held alive (which keeps the scope alive, and so the data object in the original report). You can see (one) of the paths to root here:

![image](https://user-images.githubusercontent.com/6061329/33516592-53a9acfc-d743-11e7-921a-7b0646069c0f.png)

Note that the RuleCache has a capacity of 128 so whether you run the program 128 times or 1000 times you will only have 128 "leaked" `PythonType` objects.

Anyway, all that being said, I have no idea how to go about fixing this...

**Comment by bheidkamp3** [comment]
Is this issue related to https://github.com/IronLanguages/ironpython2/issues/114 ?

**Comment by slozier** [comment]
@bheidkamp3 I don't think it's related. In fact I'm not even seeing an issue with https://github.com/IronLanguages/ironpython2/issues/114...

**Comment by bheidkamp3** [comment]
@slozier What do you mean you aren't seeing an issue with 114?
Do you think the issue has been resolved or that the initial report isn't valid?

**Comment by slozier** [comment]
I think the initial report wasn't necessarily valid (unless I'm not replicating properly). I tested with the latest bits and with 2.7.5 and it behaves the same way. The behavior I'm seeing is: first run allocations, then no new memory allocations up until the 32nd iteration where lambda generation (or something) is occurring and after that memory usage is stable for the next 99968 runs. Note that lambda generation is occurring in the background over several iterations so it might seem like memory keeps going up.
